flink-formats/flink-parquet/src/main/java/org/apache/flink/formats/parquet/ParquetInputFormat.java

```diff
@@ -168,6 +176,16 @@ public void open(FileInputSplit split) throws IOException {
         ParquetReadOptions options = ParquetReadOptions.builder().build();
         ParquetFileReader fileReader = new ParquetFileReader(inputFile, options);
         MessageType fileSchema = fileReader.getFileMetaData().getSchema();
+        if (expectedFileSchema == null) { // user did not provide a read schema, use the file schema
+            if (fieldNames == null) { // user did not provide projected fields, use all the fields
+                RowTypeInfo rowTypeInfo =
+                        (RowTypeInfo) ParquetSchemaConverter.fromParquetType(fileSchema);
+                fieldNames = rowTypeInfo.getFieldNames();
+                fieldTypes = rowTypeInfo.getFieldTypes();
+            } else { // user provided projected fields, get the corresponding types from file schema
+                fieldTypes = getFieldTypesFromSchema(fieldNames, fileSchema);
+            }
+        }
```

Comment:
:+1: you even covered the hard case with `selectFields`.

Reply:
yes :)